helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java

```diff
@@ -122,30 +122,37 @@ public void updatePreviousAssignedTasksStatus(
       Set<Integer> donePartitions = new TreeSet<>();
       for (int pId : pSet) {
         final String pName = pName(jobResource, pId);
-        TaskPartitionState currState = updateJobContextAndGetTaskCurrentState(currStateOutput,
+        TaskPartitionState currState = getTaskCurrentState(currStateOutput,
             jobResource, pId, pName, instance, jobCtx, jobTgtState);
 
-        if (!instance.equals(jobCtx.getAssignedParticipant(pId))) {
-          LOG.warn(
-              "Instance {} does not match the assigned participant for pId {} in the job context. Skipping task scheduling.",
-              instance, pId);
-          continue;
-        }
-
         // Check for pending state transitions on this (partition, instance). If there is a pending
         // state transition, we prioritize this pending state transition and set the assignment from
         // this pending state transition, essentially "waiting" until this pending message clears
+        // If there is a pending message, we should not continue to update the context because from
+        // controller prospective, state transition has not been completed yet if pending message
+        // still existed.
+        // If context gets updated here, controller might remove the job from RunTimeJobDAG which
+        // can cause the task's CurrentState not being removed when there is a pending message for
+        // that task.
         Message pendingMessage =
             currStateOutput.getPendingMessage(jobResource, new Partition(pName), instance);
-        if (pendingMessage != null && !pendingMessage.getToState().equals(currState.name())) {
```

Comment:
It looks like it's just about the order of jobContext updating vs processing pending messages, if I understood you correctly. If that's the case, is there a reason why we removed the second section of the if statement here, as @jiajunwang mentioned?

Reply:
It is very complicated to explain:
1- The controller sends RUNNINg to COMPLETED message.
2- The participant marks the current state to be COMPLETED but the pending message has not been removed yet.
3- In the next pipeline (pending message toState is COMPLETED, currentState is COMPLETED), this if statement would not be satisfied. Hence, we decide COMPLETED to DROPPED without considering pending messages.
4- The controller marks the job as completed and removes it from the DAG (so we will never consider this job again unless the controller switch to etc).
5- Since the controller sees the pending message, in the message generation phase we do not consider a new decision which is COMPLETED to DROPPED and then CS will never be DROPPED.
The conclusion is that the controller should not consider state transition as done state transition until the pending message is gone from ZK.

Reply:
Thanks for explaining. So it looks to be a problem with 2 parts:
1. Job context is updated while there are pending messages. This can cause jobs to be removed since it's marked as completed. If DROPPED message isn't correctly sent here, then the DROPPED message will never be sent because the job is deleted;
2. Current state gets updated but pending messages aren't removed (async problem). In that case, the old if statement would not catch the problem and will proceed as if the message doesn't exist. 
I suppose 2 is not a problem except when the job is about to be deleted; if the job isn't deleted, the pipeline can simply retry next time. 

For 1, I don't understand why the new code in `TaskSchedulingStage` doesn't catch it. Specifically:
```
    // Jobs that exist in current states but are missing corresponding JobConfigs or WorkflowConfigs
    // or WorkflowContexts need to be cleaned up. Note that restOfResources can only be jobs,
    // because workflow resources are created based on Configs only - workflows don't have
    // CurrentStates
    for (String resourceName : restOfResources.keySet()) {
      _workflowDispatcher.processJobForDrop(resourceName, currentStateOutput, output);
    }
```

Reply:
For 1, it is not related to the deletion of the job. The issue is one you mark the last task of the job in the jobContext as completed, then you remove this job from RunTimeJobDag in jobDispatcher. Then next pipeline you will not even see this job in your RuntimeJobDAG. Now there is no way to drop the current state until JobConfig existed. After JobConfig is removed, then the code you pasted should take care of it.